docs/proposals/20201019-prometheus-rules-in-observatorium-api.md

```diff
@@ -0,0 +1,238 @@
+* Storage permissions are managed as Kubernetes objects.
+* We could leverage Recording Rules CRDs and get validation for free (?)
+
+Cons:
+* Implicitly ties Observatorium's implementation into a Kubernetes environment. Hard to back out.
+* Large blast radius. With a large number of tenants and recording rules, we could un-intentionally impact the Kubernetes control plane's performance.
+* ETCD has a limit of 1.5MB per key / value pair so we can't store large files (but Prometheus operator seems to do this ok?).
+
+##### 2. Object Storage
+
+Pros:
+* Users of Thanos project are likely to already have object storage configured and ready to use.
+* Recording rule data model is a good fit for object storage i.e. YAML.
+
+Cons:
+* Unclear consistency semantics depending on object storage provider i.e. in the case of lots of writes, who wins?
+* Requires tenants to manage another set of secret credentials.
+
+##### 3. RDBMS
+
+Another option we have considered is to provision and use a relational database i.e. Postgres.
+
+Pros:
+* Strong isolation and update semantics.
+* Enables us to use multi-dimensional
+
+Cons:
+* We don't currently run RDBMS, and don't have any experience doing so.
+* Significant infrastructure overhead - tenants required to BYO database.
+* Requires tenants to manage another set of secret credentials.
+
+##### Decision
+
+This was discussed in the [Observatorium Community Meeting](https://docs.google.com/document/d/1jLvOH0Lllt-ShXVgWeeHBGPb3ZuagirabD_hbq41EKs/edit#heading=h.wxenjcrf4iee) on 2021-07-31.
+
+The consensus among the group was that object storage offered us the best combination of ease of use, familiarity for existing users, and also compatability with the required operations against tenants rules.
+
+We discussed that we will be likely to employ a RDBMS relatively soon for other features (i.e. RBAC), but we decided to cross that bridge when we came to it.
+
+### Sequence Diagram
+
+How will this all work in practice?
+
+![sequence-diagram](../assets/recording-rules-api.svg)
+
+#### 1. Store Tenant Rules
+1. Tenant sends request to the API using the path `/api/metrics/v1/{tenant}/rules` containing their rules data.
+2. API performs authentication and authorization of the tenant then forwards the request to the rules storage backend.
+3. Rules storage backend performs validation of the payload, rejecting any requests found to contain `alert` rules, then stores the rule file in object storage at the path `metrics/rules/{tenant}/{file_name}.yaml`.
+4. Object storage returns success to the Rule Backend.
+5. Rule backend returns success to the API.
+6. API returns success to the Tenant.
+
+#### 2. Sync Rules to Rule
+1. Periodically, thanos-ruler-syncer queries the API for all rules for all tenants.
+   * Do we need to expose an endpoint to return all of the rules? Unclear at the moment.
+   * NB: We can definitely make some optimisations here. Suggestions welcome.
```

Comment:
Why does the `thanos-ruler-syncer` require to poll the API to fetch all rules? Why not polling the `ruler-storage-backend`?

Requiring to poll the API for internal traffic looks like complicating the case w/o more benefits. Any internal client like the `thanos-rule-syncer` needs to work through authn/authz. I suggest to do so only if the syncer needs to scope queries per tenant.

Reply:
This means that if you ever change the storage of your rules then you also need to change the rule-sycners to point at new APIs. This is an indication to me that you are too tightly coupling things that should be decoupled.

Reply:
Good point :raised_hands: Calling the API is a hangover for the original design doc - it doesn't go into details why but I suspect that is so that Observatorium components themselves are designed to work in non-kubernetes environments so we can't just assume that the `syncer` can call `rule-storage-backend` directly.

The API interface will be the same between the API & `rule-storage-backend` so it doesn't matter _too_ much. It is more a question for how this is configured in our reference kubernetes deployment of Observatorium.

I'll add a comment to that effect in the doc :+1: